bitcoin/src/pow.rs

```diff
@@ -913,6 +926,81 @@ impl kani::Arbitrary for U256 {
     }
 }
 
+// TODO: this documentation is just a place holder, all the
+// docs need be rewritten when we converge on a design
+/// Error type while get a new block by height
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum GetBlockHeaderError {
+    /// Block Header not found
+    NotFound,
+    /// Generic error with custom message
+    Generic(String),
+}
+
+impl fmt::Display for GetBlockHeaderError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let msg = match self {
+            Self::Generic(error) => error.clone(),
+            Self::NotFound => String::from("Block header not found"),
+        };
+        write!(f, "{msg}")
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for GetBlockHeaderError {}
+
+// FIXME: we should move this in a better place
+/// TODO: add some docs here
+#[cfg_attr(all(test, mutate), mutate)]
+pub fn calculate_next_work_required<F>(
+    consensius_params: Params,
+    last_height: Height,
+    get_block_header_by_heigh: F,
+) -> Result<u32, GetBlockHeaderError>
+where
+    F: Fn(u32) -> Result<Header, GetBlockHeaderError>,
+{
+    let adjustment_interval = consensius_params.difficulty_adjustment_interval();
+    let pow_target_timespan = consensius_params.pow_target_timespan;
+    let no_pow_retargeting = consensius_params.no_pow_retargeting;
+    let last_adjustment_height = last_height.to_consensus_u32() as u64 - (adjustment_interval - 1);
+    let last_adjustment_block = get_block_header_by_heigh(last_adjustment_height as u32)?;
+    let last_adjustment_time = last_adjustment_block.time;
+    let current_height: u64 = last_height.to_consensus_u32() as u64;
+
+    if (current_height % adjustment_interval.min(1)) != 0 {
+        return Ok(last_adjustment_block.target().to_compact_lossy().to_consensus());
```

Comment:
If adjustment_interval is 2016 and current height is 2019, then the "next work required" here would be the target used between block 0~2015... that doesn't seem right.

Reply:
I use the same logic https://github.com/bitcoin/bitcoin/blob/master/src/pow.cpp#L19

Can you be more precise in what I can do to make this look right? thanks

Reply:
`pindexLast` in bitcoin core is the equivalent of the block at `current_height` in your code.

However, `last_adjustment_block` is the block at `current_height - 2015`

The variable names in bitcoin core's code are confusing, and your code also uses confusing names that are trying to mimic core's names.

Perhaps you should rename all the variables to make things easier to follow.